### ISSUE TITLE
[FEATURE] Improve the abstraction around ztoc

### ISSUE THREAD

**Issue by djdongjin** [awslabs/soci-snapshotter#463]
### Description

The current structure of ztoc&compression has some issues (see discussion in https://github.com/awslabs/soci-snapshotter/pull/453#discussion_r1122174133):

1. Zinfo is part of ztoc, compression is used to implement zinfo. Thus, compression pkg should be a sub-package of ztoc package.
2. Also, zinfo should be always part of a ztoc object. As an example, spanmanager [holds both ztoc and zinfo objects](https://github.com/awslabs/soci-snapshotter/blob/00474ecf61cfb905a0d6436b516110050bb6435c/fs/span-manager/span_manager.go#L122-L125).
3. Ztoc struct should have the zinfo **struct**, not its [serilized **bytes**](https://github.com/awslabs/soci-snapshotter/blob/878e7d9fbdb67fd030f8ac1b36777a83cd525aea/ztoc/ztoc.go#L59). We should only serilize zinfo when we serilize the whole ztoc.
4. Ztoc/Zinfo should have the compression algorithm used to produce that ztoc/zinfo. This info is already in [ztoc's flatbuf](https://github.com/awslabs/soci-snapshotter/blob/878e7d9fbdb67fd030f8ac1b36777a83cd525aea/ztoc/fbs/ztoc.fbs#LL31C8-L31C8)

### Describe the solution you'd like

Resolve the above issues (or more if found).

### Describe any alternative solutions/features you've considered

Don't change anything.

**Comment by Kern--** [comment]
Copying this discussion out of the PR so we can continue it here.

---
> What direction are we going with this API?
> 
> Are we going to keep embedding `CompressionInfo` into `Ztoc` with this helper? Should we instead embed `ZInfo` into `Ztoc` and just use the ztoc directly instead of `ztoc.Zinfo()`?

I tried embedding `Zinfo` into `Ztoc` but it's a little tricky due to the gzipzinfo c resource release. In some places it will cause the Zinfo creation and release done in different places in some scenarios. https://github.com/awslabs/soci-snapshotter/issues/463#issuecomment-1478335134 has more details.

After digging into the issue, I think both ways have their pros and cons:

Keep zinfo struct in ztoc:

pros:

- simplify zinfo usage: just `ztoc.Zinfo`.

cons:

- whenever a ztoc is created, its zinfo needs to be closed to release the c resource (ideally by the same creator).
- when to close the zinfo might be tricky, especially the ztoc is passed to a func as an arg (should caller or callee release?).

Keep zinfo bytes in ztoc (current approach, my preference while we still have c resource):

pros:

- sometimes only the TOC part of the ztoc is used, not the Zinfo part. So don't need to take care of Zinfo release.
- if someone needs to use Zinfo, it can get one by calling `ztoc.Zinfo()` and release after its done (more clear on who should release).

cons:

- Need an extra step to convert bytes to struct (i.e., `ztoc.Zinfo()`).

_Originally posted by @djdongjin in https://github.com/awslabs/soci-snapshotter/issues/513#issuecomment-1478428779_

**Comment by Kern--** [comment]
If we embed zinfo, then `Ztoc` _becomes_ a `ZInfo`. You can do `ztoc.ExtractFile()` just like you can do `ztoc.FileMetadata[0]`. 

If we don't embed the zinfo, then we don't solve the problem that `SpanManager` has both a ztoc and a zinfo. 


I think the real root of the problem is that we don't have a clear idea of ownership. We've been trying to move towards a model where everything is nicely wrapped up in the ztoc, but we blow that up by giving the TOC to metadata and the span manager reimplements the ztoc extraction logic so it can do fancier data management. We can make these small decisions to get something to work, but I think we're always going to have this tension if we don't solve the ownership problem.

**Comment by djdongjin** [comment]
> If we embed zinfo, then Ztoc becomes a ZInfo. You can do ztoc.ExtractFile() just like you can do ztoc.FileMetadata[0].

Yes, embedding `Zinfo` into `Ztoc` makes things easier (especially when `Zinfo` is used).

> If we don't embed the zinfo, then we don't solve the problem that SpanManager has both a ztoc and a zinfo.

Yes, the issue is not resolved. #513 just makes some existing code simplier/cleaner.

> I think the real root of the problem is that we don't have a clear idea of ownership. We've been trying to move towards a model where everything is nicely wrapped up in the ztoc, but we blow that up by giving the TOC to metadata and the span manager reimplements the ztoc extraction logic so it can do fancier data management. We can make these small decisions to get something to work, but I think we're always going to have this tension if we don't solve the ownership problem.

I agree. There is definitely some work to be done to solve this. Meanwhile, I feel to some extent the `ownership problem` is brought up by the fact that we have c resource that needs to be released (imagine a pure go impl, a `Ztoc` may just be a normal go struct that every one can use as long as it exists and will be destructed when out of its scope). So to resolve the issue:

1. Removing c resource usage may help resolve this.
2. As you said, better define the ownership of ztoc.

I haven't had other thoughts on this. Will update in the issue if I figure something out.

**Comment by Kern--** [comment]
Blocked on either https://github.com/awslabs/soci-snapshotter/issues/76 or defining the ownership to make sure that zinfo is properly closed.
